File: tests/tasks-sasl-plaintext/src/test/java/docs/MyTest.java
package docs;

// tag::imports[]
import io.micronaut.configuration.kafka.annotation.*;
import io.micronaut.test.extensions.junit5.annotation.MicronautTest;
import jakarta.inject.Inject;
import org.junit.jupiter.api.Test;

import static java.util.concurrent.TimeUnit.SECONDS;
import static org.awaitility.Awaitility.await;
// end::imports[]

// tag::clazz[]
@MicronautTest
class MyTest extends AbstractKafkaTest {

    @Inject
    MyProducer producer;
    @Inject
    MyConsumer consumer;

    @Test
    void testKafkaRunning() {
        // Given
        final String message = "hello";
        // When
        producer.produce(message);
        // Then
        await().atMost(5, SECONDS).until(() -> message.equals(consumer.consumed));
    }

    @KafkaClient
    public interface MyProducer {
        @Topic("my-topic")
        void produce(String message);
    }

    @KafkaListener(offsetReset = OffsetReset.EARLIEST)
    public static class MyConsumer {
        String consumed;
        @Topic("my-topic")
        public void consume(String message) {
            consumed = message;
        }
    }
}
// end::imports[]

File: test-suite-python/src/test/python/micronaut/kafka/docs/MyTest.py
from abc import ABC, abstractmethod
from queue import Empty, Queue
from typing import Annotated

from jakarta.inject import Inject
from micronaut.configuration.kafka.annotation import KafkaClient, KafkaListener, OffsetReset, Topic
from micronaut.context.annotation import Property, Requires
from micronaut.test.extensions.junit5.annotation import MicronautTest
from org.junit.jupiter.api import Disabled, Test, TestInstance

from .AbstractKafkaTest import AbstractKafkaTest


@Requires(property="spec.name", value="MyTest")
@KafkaClient
class MyProducer(ABC):

    @Topic("my-topic")
    @abstractmethod
    def produce(self, message: str) -> None:
        ...


@Requires(property="spec.name", value="MyTest")
@KafkaListener(offsetReset=OffsetReset.EARLIEST)
class MyConsumer:

    def __init__(self):
        self.consumed_messages: Queue[str] = Queue()

    @Topic("my-topic")
    def consume(self, message: str) -> None:
        self.consumed_messages.put(message)

    def await_message(self, timeout: float) -> str | None:
        try:
            return self.consumed_messages.get(timeout=timeout)
        except Empty:
            return None


# TODO(python): Micronaut Test calls TestPropertyProvider.getProperties() before the application
# context, and with it the GraalPy runtime, exists ("GraalPy context has not been initialized"), so a
# Python test class cannot supply the container properties yet. The other Python tests of this suite
# get the bootstrap servers from the Java KafkaTestConfigurer of the "kafka" environment instead.
@Disabled("TODO(python): TestPropertyProvider.getProperties() runs before the GraalPy runtime exists")
@Property(name="spec.name", value="MyTest")
@MicronautTest
@TestInstance(TestInstance.Lifecycle.PER_CLASS)
class MyTest(AbstractKafkaTest):
    producer: Annotated[MyProducer, Inject]
    consumer: Annotated[MyConsumer, Inject]

    @Test
    def test_kafka_running(self):
        message = "hello"
        self.producer.produce(message)
        assert self.consumer.await_message(15) == message
